tests/bin/difference.py

```diff
@@ -0,0 +1,420 @@
+
+################################################################
+
+def run_viewer(proof_root, viewer1, root1, viewer2, root2, litani):
+    logging.info("Loading and updating viewer jobs...")
+    viewer_jobs = viewer_add_job_commands(proof_root)
+    jobs = update_jobs(viewer_jobs, viewer1, root1) + update_jobs(viewer_jobs, viewer2, root2)
+
+    logging.info("Running litani init...")
+    run([litani, "init", "--project", "Differential testing"])
+
+    for job in jobs:
+        cmd = [litani, "add-job"]
+        for flag, value in job.items():
+            cmd.append(flag)
+            if isinstance(value, bool):
+                continue
+            if isinstance(value, list):
+                cmd.extend(value)
+            else:
+                cmd.append(value)
+        logging.info("Running litani add-job...")
+        run(cmd)
+
+    logging.info("Running litani run-build...")
+    run([litani, "run-build"], capture_output=False)
+
+################################################################
+
+def compare_reports(reports1, reports2):
+    logging.info("Comparing %s and %s...", reports1, reports2)
+    try:
+        run(["diff", "-r", reports1, reports2])
+    except subprocess.CalledProcessError as error:
+        raise UserWarning("Reports differ: compare with 'diff -r {reports1} {reports2}'") from error
+
+################################################################
+# Validate command line arguments
+
+def check_git_status(repo):
+    try:
+        status = run(["git", "status", "--porcelain"], cwd=repo, capture_output=True)
+    except subprocess.CalledProcessError as error:
+        raise UserWarning("This is not a git repository: git status failed") from error
+    if status:
+        raise UserWarning("This git repository has uncomitted changes.")
+    return True
+
+def validate_litani(litani):
+    logging.debug("--litani = %s", litani)
+    try:
+        return run(["which", litani], capture_output=True)[0]
+    except subprocess.CalledProcessError as error:
+        raise UserWarning("Use --litani to give the command invoking litani") from error
+
+def validate_repository(repo):
+    logging.debug("--viewer-repository = %s", repo)
+    try:
+        assert isinstance(repo, str)
+        assert os.path.exists(f"{repo}{os.sep}.git")
+        return repo
+    except AssertionError as error:
+        raise UserWarning(
+            "Use --viewer-repository to give the root of the cbmc-viewer repository"
+        ) from error
+
+def validate_proofs(proofs):
+    logging.debug("--proofs = %s", proofs)
+    try:
+        assert isinstance(proofs, str)
+        assert os.path.exists(f"{proofs}{os.sep}.litani_cache_dir")
+        return proofs
+    except AssertionError as error:
+        raise UserWarning("Use --proofs to specify a proof directory "
+                          "(containing .litani_cache_dir)") from error
+
+def validate_commits(commits):
```

Comment:
Suggest to actually validate them here, e.g. for each of the commits, `git rev-parse commit`? If the user mis-spelled `HEAD` or a hash it would be nice to fail here.

Reply:
Done.  Validation code made more precise and simpler.